File: 1052-grumpy-bookstore-owner/1052-grumpy-bookstore-owner.java
class Solution {
    public int maxSatisfied(int[] customers, int[] grumpy, int minutes) {
        int N = customers.length;
        
        int[] oriCount = new int[N];
        int[] teqCount = new int[N];
        int[] accumulated = new int[N];
        accumulated[0] = customers[0];
        for(int i = 1; i < N; i++) {
            accumulated[i] = accumulated[i - 1] + customers[i];
        }
        
        /* 0부터 시작 */
        oriCount[0] = (grumpy[0] + 1) % 2 * customers[0];
        for(int i = 0; i < minutes; i++) {
            teqCount[minutes - 1] += customers[i];
        }

        for(int i = 1; i < N; i++) { //초마다 순회
            oriCount[i] = oriCount[i - 1] + (grumpy[i] + 1) % 2 * customers[i];
            teqCount[i] = Math.max(teqCount[i], teqCount[i - 1] + (grumpy[i] + 1) % 2 * customers[i]);
            
            if(i + minutes <= N)
                teqCount[i + minutes - 1] = Math.max(
                    teqCount[i + minutes - 1],
                    oriCount[i - 1] + accumulated[i + minutes - 1] - accumulated[i - 1]
                );
        }
        
        // System.out.printf("짜증 그대로 둘 때의 Customer Count: %s\n", Arrays.toString(oriCount));
        // System.out.printf("짜증 사용할 때의 Customer Count: %s\n", Arrays.toString(teqCount));
        
        return teqCount[N - 1];
    }
}

File: 1052-grumpy-bookstore-owner/1052-grumpy-bookstore-owner.py
class Solution:
    def maxSatisfied(self, customers: List[int], grumpy: List[int], minutes: int) -> int:
        N = len(customers)
        oriCount = [0] * N
        teqCount = [0] * N
        accumulated = list()
        tmp = 0
        for i in customers:
            tmp += i
            accumulated.append(tmp)
        
        oriCount[0] = customers[0] * (grumpy[0] ^ 1)
        teqCount[minutes - 1] = sum(customers[:minutes])
    
        
        for i in range(1, N):
            oriCount[i] = oriCount[i - 1] + customers[i] * (grumpy[i] ^ 1)
            teqCount[i] = max(teqCount[i], teqCount[i - 1] + customers[i] * (grumpy[i] ^ 1))
            
            if i + minutes <= N:
                teqCount[i + minutes - 1] = max(
                    teqCount[i + minutes - 1],
                    oriCount[i - 1] + accumulated[i + minutes - 1] - accumulated[i - 1]
                )
        
        
        return teqCount[N-1]